fix(ngUpgrade): downgrade projection

petebacondarwin opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
[x] Feature
[x] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

This is just a re-jigging of #9538 - @youdz please take a look

petebacondarwin committed:
refactor(compiler/template_parser): export `createElementCssSelector`

This is needed in `ngUpgrade`.

petebacondarwin committed:
refactor(upgrade/util): remove unused `stringify()` method

petebacondarwin committed:
refactor(upgrade): add missing Angular 1 type info

youdz committed:
fix(upgrade): fix downgrade content projection and injector inheritance

- Full support for content projection in downgraded Angular 2
  components. In particular, this enables multi-slot projection and
  other features on <ng-content>.
- Correctly wire up hierarchical injectors for downgraded Angular 2
  components: downgraded components inherit the injector of the first
  other downgraded Angular 2 component they find up the DOM tree.

Closes #6629, #7727, #8729, #9643, #9649, #12675

petebacondarwin committed:
docs(upgrade/upgrade_adapter): fix up references to AngularJS and Angular 2

petebacondarwin commented:
@youdz - I made a few API changes. I hope these fit with you.
In particular, I made the `declareNg1Module` private for now - it is not really needed publicly and it is harder to make things private later than the other way around.
I also renamed some stuff...
* `PromiseCompleter` -> `Deferred`
* `initNg1ModuleForTests` -> `registerNg1ModuleForTests`
* the "spyIng" stuff becomes: `CapturedContentSelectors` and `RuntimeCompilerSpyFactory`

I broke the commits up into smaller pieces. It might be worth reviewing commit by commit.
The two main commits I kept with you as the author. Please check that you are happy with that.

mhevery commented:
This deals with compiler, and so I think @tbosch should have a look.

youdz commented:
Awesome @petebacondarwin, I can't take a look right now but I'll review first thing in the morning tomorrow. I'm on the East Coast, so hopefully I'll be done before you even start working tomorrow.

petebacondarwin commented:
Except I'm in UK :-)

On 3 Nov 2016 00:23, "Eudes Petonnet-Vincent" <notifications@github.com>
wrote:

> Awesome @petebacondarwin <https://github.com/petebacondarwin>, I can't
> take a look right now but I'll review first thing in the morning tomorrow.
> I'm on the East Coast, so hopefully I'll be done before you even start
> working tomorrow.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/12675#issuecomment-258038844>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA9JyboKpIOfZsNPOqT9OxZ-1l-fBlNks5q6SmGgaJpZM4Kn2bj>
> .
>

petebacondarwin commented on tools/public_api_guard/upgrade/index.d.ts:
```diff
@@ -1,3 +1,6 @@
+/** @experimental */
+export declare const platformUpgrade: (extraProviders?: Provider[]) => PlatformRef;
```
@misko Is there some way so that this doesn't have to be in the public API? I think the only reason we need this here is that we want to access it in our unit tests.

petebacondarwin commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -505,58 +398,274 @@ export class UpgradeAdapter {
+            (<any>element[0]).id = id;
+
+            let injectorPromise = new ParentInjectorPromise();
+            element.data(controllerKey(NG2_INJECTOR), injectorPromise);
```
It just occurred to me that we shouldn't be storing a promise under the NG2_INJECTOR key since people might rely upon this being the actual ng2 injector...

gkalpak commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -507,64 +451,289 @@ export class UpgradeAdapter {
+  resolve(injector: Injector) {
+    this.injector = injector;
+
+    // reset the element data to poin to the real injector
```
_poin_ -->  _point_

gkalpak commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -507,64 +451,289 @@ export class UpgradeAdapter {
       parse: angular.IParseService): angular.IDirective {
-    var idCount = 0;
+    let idCount = 0;
+    let dashSelector = info.selector.replace(/([A-Z])/g, char => '-' + char.toLowerCase());
```
The parenthesis around `[A-Z]` is unnecessary.

gkalpak commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -507,64 +451,289 @@ export class UpgradeAdapter {
+          linkFns = compileProjectedNodes(templateElement);
         }
+        return {
+          post: (scope: angular.IScope, element: angular.IAugmentedJQuery,
```
Nit: You can return the post-link function directly (without the `{post: ...}` object).

petebacondarwin replied:
Yes but I have a feeling that it was done this way for a reason... so we are copying what was already there.

youdz commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -272,6 +272,59 @@ export class UpgradeAdapter {
+    if (!isPresent(windowNgMock) || !isPresent(windowNgMock.module)) {
+      throw new Error('Failed to find \'angular.mock.module\'.');
+    }
+    this.declareNg1Module(modules);
```
If I `registerForNg1Tests()`  in a `beforeEach()`, I'd end up re-declaring the upgrader's Angular 1 module for each test. Isn't it better to declare the module once, then call this to only re-run the configuration code every time, not the whole declaration?

If that's the case, then we do need `declareNg1Module()` to be part of the public API.

petebacondarwin replied:
I thought about this for a while but I think the following two points are valid:
* I don't believe that re-declaring is a big performance cost, we do it all the time in our Angular 1 tests. But I would be happy to back down on that if you show me the numbers.
* the trouble is that there is a lot of state being held in the adapter, such as `this.ngZone` and `this.ng2BootstrapDeferred` which really needs reinitializing on each test. We could do a big refactor to make `declareNg1Module` "stateless" but as it is right now I think it is safer to call it every time.

youdz replied:
Sounds good to me, let's stick to redeclaring then. Better safe than fast.

youdz commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -272,6 +272,59 @@ export class UpgradeAdapter {
+    windowNgMock.module(this.ng1Module.name);
+    const upgrade = new UpgradeAdapterRef();
+    this.ng2BootstrapDeferred.promise.then(
+        () => { (<any>upgrade)._bootstrapDone(this.moduleRef, this.moduleRef); }, onError);
```
`(<any>upgrade)._bootstrapDone(this.moduleRef, this.ng1Injector)`

gkalpak replied:
Doesn't this need fixing too. BTW, I don't think `this.ng1Injector` works here, because `this.ng1Injector` is set inside `_bootstrapDone`. But now the `ng1Injector` is available as the resolution value for the `ng2BootstrapDeferred.promise`. So, this can be changed to:

```ts
this.ng2BootstrapDeferred.promise.then(
          ng1Injector => { (<any>upgrade)._bootstrapDone(this.moduleRef, ng1Injector); }, onError);
```

(But why isn't any test failing? :confused)

youdz reviewed (commented):
Thanks for the clean-up, especially the comments. When I saw that the currently existing platforms were not commented at all, I figured I'd just follow this. 
But clearly too many comments would have been better than not enough. :-P

youdz commented on modules/@angular/upgrade/src/platform_upgrade.ts:
```diff
@@ -0,0 +1,24 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {CompilerFactory, PlatformRef, Provider, createPlatformFactory} from '@angular/core';
```
Actually it is, the exported `platformUpgrade` is a PlatformRef, so we need to import the type to be able to export it, even if we don't write it explicitly.

gkalpak replied:
I didn't understand what you just said, but I am sure you known what you are talking about :smiley:

petebacondarwin replied:
Ahah! Now I look again, I agree with @gkalpak - we don't actually use this import

youdz commented:
@googlebot: I'm okay with my commits being contributed to Google.

It would be a pretty bad joke otherwise.

petebacondarwin commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -507,64 +451,289 @@ export class UpgradeAdapter {
+    this.ng2BootstrapDeferred = new Deferred();
+    ng1Module.factory(NG2_INJECTOR, () => this.moduleRef.injector.get(Injector))
+        .constant(NG2_ZONE, this.ngZone)
+        .constant(NG2_COMPONENT_FACTORY_REF_MAP, componentFactoryRefMap)
```
I wanted for them to be available to providers...

gkalpak replied:
That's a noble goal indeed :smiley: It comes at the cost of making them "un-decorate-able", of course. No idea which usecase is more useful/common, but I trust you have thought this through.

petebacondarwin replied:
Undecoratable true - I can't think of a use case for that. You can still override them...

petebacondarwin commented on modules/@angular/upgrade/src/runtime_compiler_spy.ts:
```diff
@@ -0,0 +1,56 @@
+
+import {RuntimeCompilerFactory} from '@angular/compiler';
+import {COMPILER_OPTIONS, Compiler, CompilerOptions, Inject, Injectable} from '@angular/core';
+import {isPresent} from './facade/lang';
```
Well we have brought the facade folder into the upgrade project for AoT so it is there.
The real question is whether we actually need to use `!isPresent(...)` at all rather than just `!...`

mhevery replied:
Yes please remove all facades as we are actively working on deleting them. Let's not create new ones.

petebacondarwin committed:
refactor(upgrade/upgrade_adapter): use `Deferred` helper

Making Angular 1's `$compile` asynchronous by chaining injector promises
in linking functions can cause flickering views in applications.

mhevery commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -7,24 +7,26 @@
-import {NG1_COMPILE, NG1_INJECTOR, NG1_PARSE, NG1_ROOT_SCOPE, NG1_TESTABILITY, NG2_COMPILER, NG2_COMPONENT_FACTORY_REF_MAP, NG2_INJECTOR, NG2_ZONE, REQUIRE_INJECTOR} from './constants';
+import {NG1_COMPILE, NG1_INJECTOR, NG1_PARSE, NG1_ROOT_SCOPE, NG1_TESTABILITY, NG2_CAPTURED_CONTENT_SELECTORS, NG2_COMPILER, NG2_COMPONENT_FACTORY_REF_MAP, NG2_INJECTOR, NG2_ZONE, REQUIRE_INJECTOR} from './constants';
 import {DowngradeNg2ComponentAdapter} from './downgrade_ng2_adapter';
+import {isPresent} from './facade/lang';
```
remove here as well

mhevery commented on modules/@angular/upgrade/src/util.ts:
```diff
@@ -5,12 +5,8 @@
-  return '' + obj;
-}
+import {CssSelector, SelectorMatcher, createElementCssSelector} from '@angular/compiler';
+import {isPresent} from './facade/lang';
```
remove

mhevery commented:
My only comment is to remove all facades, LGTM otherwise

youdz committed:
feat(upgrade): enable Angular 1 unit testing of upgrade module

- New method `UpgradeAdapter.registerForNg1Tests(modules)` declares the
  Angular 1 upgrade module and provides it to the `angular.mock.module()`
  helper.
  This prevents the need to bootstrap the entire hybrid for every test.

Closes #5462, #12675

petebacondarwin commented:
CI is green and I have rebased on master.
@mhevery happy to merge?

gkalpak commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -269,17 +272,70 @@ export class UpgradeAdapter {
+   * let $compile, $rootScope;
+   *
+   * beforeEach(() => {
+   *   upgradeAdapterRef = upgradeAdapter.registerForNg1Tests(['heroApp'])
```
Now that you mention it, does it mean that you can use it like:

```js
beforeEach(upgradeAdapter.registerForNg1Tests(['heroApp']))
// Similar to:
// beforeEach(module('heroApp'))
```

If not, then the above comment can be misleading :smiley:

petebacondarwin replied:
Oh you spoilsport! OK, we need to reword

youdz commented:
@petebacondarwin, @mhevery: Any update on this?

I'd be happy to help, but I'm not sure what to do now that Pete moved to his branch.

petebacondarwin commented:
@youdz - I'm trying to squeeze it into 2.2.0 which is imminent. If not it might have to wait for 2.3.0

petebacondarwin commented:
Looks like 2.2.0 is already at RC and so this can't land there. We will get it into 2.3.0 which will be out in a month

youdz commented:
Hi there, so now that 2.2.0 is out, what can I do to make sure this makes it into 2.3.0? We're trying to align some our timelines based on this, so we really don't want to miss it.

@petebacondarwin: Do you mind if I help by pushing short fixes to your branch, for instance?

petebacondarwin commented:
I'll rebase and ping @mhevery for a review. 
What small fixes did you have in mind?

youdz commented:
The build failed because of golden files differences, so I was thinking of that kind of fixes. One or two lines in golden files, maybe one or two if the linter failed, things like that. Nothing too specific for now, I was just asking in general.

petebacondarwin commented:
It seems that this PR has become stale as the compiler has changed :-(
I'll try a rebase/reworking on Monday...

@tbosch you had some suggestion of how to do this without monkeypatching the compiler?

tbosch commented on modules/@angular/upgrade/src/platform_upgrade.ts:
```diff
@@ -0,0 +1,24 @@
+ *
+ * @experimental
+ */
+export const platformUpgrade =
```
This is only for patching the compiler, right? If we didn't need to patch the compiler, and the compiler had a method that can just give you the content selectors, this would not be needed?

petebacondarwin replied:
Yes that is right.

tbosch commented on modules/@angular/upgrade/src/runtime_compiler_spy.ts:
```diff
@@ -0,0 +1,55 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {RuntimeCompilerFactory} from '@angular/compiler';
```
Let's change the `RuntimeCompiler` to have a method `getNgContentSelectors(component: Type<any>)` that uses a cache that is filled during compilation...

Then we don't need this hack...

tbosch replied:
See https://github.com/angular/angular/blob/master/modules/@angular/compiler/src/jit/compiler.ts#L158, where we have the `CompileDirectiveMetadata` which contains the `CompileTemplateMetadata` which contains the `ngContentSelectors`.

tbosch replied:
So you could add this method to the `RuntimeCompiler`:
```
getNgContentSelectors(component: Type<any>): string[] {
  const template = this._compiledTemplateCache.get(component);
  if (!template) {
    throw new Error(`The component ${stringify(component)} is not yet compiled!`);
  }
  return template.compMeta.template.ngContentSelectors
}
```

petebacondarwin replied:
OK. I will work on that tomorrow. Thanks!

jaffoneh commented:
hey @petebacondarwin: since this is a change we're depending on moving forward, I was wondering if there is an estimate of when will this make it to Angular 2. I see 2.2.4 was released today and I was wondering if this is going to make into 2.2.5 or if there is a way we could provide any help to accelerate it? Thanks again for your help! cc @youdz

petebacondarwin commented:
@jaffoneh - this should be ready now. I had to do a bit of refactoring so we were not monkeypatching the compiler.

@tbosch - can you please review the compiler changes?

tbosch commented on modules/@angular/core/src/linker/compiler.ts:
```diff
@@ -82,6 +82,8 @@ export class Compiler {
     throw _throwError();
   }
 
+  getNgContentSelectors(component: Type<any>): string[] { throw _throwError(); }
```
Could you add a comment describing the purpose and when this can be called? (e.g. that the component has to be compiled...)

petebacondarwin replied:
will do!

petebacondarwin replied:
done

tbosch commented on modules/@angular/upgrade/src/util.ts:
```diff
@@ -27,3 +22,59 @@ export function onError(e: any) {
+  return asArray || [];
+}
+
+export function sortProjectableNodes(ngContentSelectors: string[], childNodes: Node[]): Node[][] {
```
I think this function deserves some unit tests for all the corner cases (e.g. combination of wildcard and non wildcard ng-content, ...).

petebacondarwin replied:
OK

petebacondarwin replied:
done

tbosch commented on modules/@angular/upgrade/src/util.ts:
```diff
@@ -27,3 +22,59 @@ export function onError(e: any) {
+}
+
+
+export class Deferred<R> {
```
This class seems to be unused...

tbosch commented on tools/public_api_guard/upgrade/index.d.ts:
```diff
@@ -4,6 +4,7 @@ export declare class UpgradeAdapter {
     bootstrap(element: Element, modules?: any[], config?: angular.IAngularBootstrapConfig): UpgradeAdapterRef;
     downgradeNg2Component(type: Type<any>): Function;
     downgradeNg2Provider(token: any): Function;
+    registerForNg1Tests(modules?: string[]): UpgradeAdapterRef;
```
Why is this line added? There is no code that refers to it in this PR...

tbosch replied:
Ah, never mind. Github did not expand all files...

tbosch commented:
LGTM for everything except modules/@angular/upgrade/src/upgrade_adapter.ts (I don't know the details in there...)

gkalpak commented on modules/@angular/upgrade/test/upgrade_spec.ts:
```diff
@@ -178,7 +178,7 @@ export function main() {
              expect(document.body.textContent).toEqual('1A;2A;ng1a;2B;ng1b;2C;1C;');
              // https://github.com/angular/angular.js/issues/12983
-             expect(log).toEqual(['1A', '1B', '1C', '2A', '2B', '2C', 'ng1a', 'ng1b']);
+             expect(log).toEqual(['1A', '1C', '2A', '2B', '2C', 'ng1a', 'ng1b']);
```
Why this change?

gkalpak commented on modules/@angular/upgrade/test/util_spec.ts:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    describe('sortProjectableNodes', () => {
+      it('should should return an array of node collections for each selector', () => {
```
_should should_ --> _should_

gkalpak reviewed (commented):
Basically LGTM.
I think there should be some tests for `registerForNg1Tests()`, though.

petebacondarwin commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -271,6 +270,60 @@ export class UpgradeAdapter {
+   * run assertions once the Angular 2+ components are ready to test through Angular 1.
+   */
+  registerForNg1Tests(modules?: string[]): UpgradeAdapterRef {
+    const windowNgMock = (window as any)['angular'].mock;
```
yes - since the "official" type for `window` does not include index access.

petebacondarwin commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
````diff
@@ -269,17 +270,71 @@ export class UpgradeAdapter {
+   * ### Example
+   *
+   * ```
+   * const upgradeAdapter = new UpgradeAdapter();
````
OK

petebacondarwin committed:
docs(upgrade): fix UpgradeAdapter examples

vicb commented:
Fails in G3
`/modules/@angular/upgrade/src/util.ts(8,70): error TS2307: Cannot find module '@angular/compiler'`

alexeagle commented:
@petebacondarwin we get that error because upgrade cannot depend on compiler. (We have the static-only upgrade)

I think we need a util-static.ts which is included in our library, and a util.ts which is not...

petebacondarwin commented:
Thanks for looking at this @alexeagle.

So I can put the bits that rely upon the JIT compiler into a separate module, which is not imported by any of the `upgrade/static` modules. Would that solve the problem?

alexeagle commented:
It doesn't have to be a separate module, we just have the following code that the caretaker might need to modify when you change things:
@vicb you should probably try to fix this yourself since the PR comes from non-googler and it's our own internal build that needs to see the right file subsets....

```
UPGRADE_STATIC_SRCS = [
    "rc/modules/@angular/upgrade/src/aot/**/*.ts",
    "rc/modules/@angular/upgrade/src/facade/**/*.ts",
    "rc/modules/@angular/upgrade/src/angular_js.ts",
    "rc/modules/@angular/upgrade/src/util.ts",
    "rc/modules/@angular/upgrade/static.ts",
]

ng_module(
    name = "upgrade",
    srcs = glob(UPGRADE_STATIC_SRCS),
    deps = [
        ":angular2_es6_subset",
        ":core",
        ":platform_browser",
        "//third_party/javascript/node_modules/rxjs:canary",
    ],
)

# ngUpgrade that depends on compiler
ng_module(
    name = "upgrade_dynamic",
    srcs = glob(
        ["rc/modules/@angular/upgrade/**/*.ts"],
        exclude = UPGRADE_STATIC_SRCS + [
            "rc/modules/@angular/upgrade/testing.ts",
            "rc/modules/@angular/upgrade/testing/**",
            "rc/modules/@angular/upgrade/test/**",
        ],
    ),
    deps = [
        ":angular2_es6_subset",
        ":compiler",
        ":core",
        ":platform_browser",
        ":platform_browser_dynamic",
        ":upgrade",
        "//third_party/javascript/node_modules/rxjs:canary",
    ],
)
```

(We don't let anyone new start using the "upgrade_dynamic" module)

petebacondarwin commented:
Right, so if I split out the bits from `"rc/modules/@angular/upgrade/src/util.ts",` that depend upon the compiler into a new file (i.e. module) then it shouldn't break the static stuff.

petebacondarwin commented:
@alexeagle and @vicb I have added a commit that extracts the JIT compiler dependent bits into the `upgrade_adapter.ts` file, so it should not cause problems with `upgrade/static` any more.

Please test and squash this new commit into the one with the similar commit message, if you are happy.

youdz commented:
Thanks so much for getting this through even with all the extra work it needed to keep updating it to the latest structure and standards. 🎉

petebacondarwin commented:
Sorry it took so long @youdz

jensbodal commented:
For those like me that stumble across here, this is included in `4.0.0` but not `2.4.9`.

http://plnkr.co/edit/6SFy7nbK6TgtfQ44tMfs?p=preview

Change `ngVer` in `systemjs.config.js` to `4.0.0` to see it work.
